src/ir_Haier.cpp

```diff
@@ -654,27 +654,101 @@ void IRHaierAC176::setMode(uint8_t mode) {
 /// @return The current operating mode setting.
 uint8_t IRHaierAC176::getMode(void) const { return _.Mode; }
 
-/// Set the temperature.
-/// @param[in] celsius The temperature in degrees celsius.
-void IRHaierAC176::setTemp(const uint8_t celsius) {
-  uint8_t temp = celsius;
-  if (temp < kHaierAcMinTemp)
-    temp = kHaierAcMinTemp;
-  else if (temp > kHaierAcMaxTemp)
-    temp = kHaierAcMaxTemp;
+/// Set the default temperature units to use.
+/// @param[in] on Use Fahrenheit as the units.
+///   true is Fahrenheit, false is Celsius.
+void IRHaierAC176::setUseFahrenheit(const bool on) { _.UseFahrenheit = on; }
 
-  uint8_t old_temp = getTemp();
-  if (old_temp == temp) return;
-  if (old_temp > temp)
-    _.Button = kHaierAcYrw02ButtonTempDown;
-  else
-    _.Button = kHaierAcYrw02ButtonTempUp;
-  _.Temp = temp - kHaierAcMinTemp;
+/// Get the default temperature units in use.
+/// @return true is Fahrenheit, false is Celsius.
+bool IRHaierAC176::getUseFahrenheit(void) const { return _.UseFahrenheit; }
+
+/// Set the temperature.
+/// @param[in] degree The temperature in degrees.
+/// @param[in] fahrenheit Use units of Fahrenheit and set that as units used.
+void IRHaierAC176::setTemp(const uint8_t degree, const bool fahrenheit) {
+  if (fahrenheit) {
+    uint8_t temp = degree;
+    if (temp < kHaierAcYrw02MinTempF)
+      temp = kHaierAcYrw02MinTempF;
+    else if (temp > kHaierAcYrw02MaxTempF)
+      temp = kHaierAcYrw02MaxTempF;
+
+    uint8_t old_temp = getTemp();
+    if (old_temp == temp) return;
+    if (old_temp > temp)
+      _.Button = kHaierAcYrw02ButtonTempDown;
+    else
+      _.Button = kHaierAcYrw02ButtonTempUp;
+
+    if (degree >= 77) { temp++; }
+    if (degree >= 79) { temp++; }
+    // See at IRHaierAC176::getTemp() comments for clarification
+    _.ExtraDegreeF = temp % 2;
+    _.Temp = (temp - kHaierAcYrw02MinTempF -_.ExtraDegreeF) >> 1;
+    _.UseFahrenheit = true;
+  } else {
+    uint8_t temp = degree;
+    if (temp < kHaierAcYrw02MinTempC)
+      temp = kHaierAcYrw02MinTempC;
+    else if (temp > kHaierAcYrw02MaxTempC)
+      temp = kHaierAcYrw02MaxTempC;
+
+    uint8_t old_temp = getTemp();
+    if (old_temp == temp) return;
+    if (old_temp > temp)
+      _.Button = kHaierAcYrw02ButtonTempDown;
+    else
+      _.Button = kHaierAcYrw02ButtonTempUp;
+    _.Temp = temp - kHaierAcYrw02MinTempC;
+    _.UseFahrenheit = false;
+  }
 }
```

Comment:
There is a LOT of duplicate code in the `if (fahrenheit)` clauses. As we try to code to reduce the space used, de-duping this will save space. It also is better code style too.
For example, You can put this at the end, outside the `if` statements.
```c++
_.UseFahrenheit = fahrenheit;
```
and the button logic can go before the `if` statement too (these are not the only reductions/de-duplications)
```c++
  uint8_t old_temp = getTemp();
  if (old_temp == temp) return;
  if (old_temp > temp)
    _.Button = kHaierAcYrw02ButtonTempDown;
  else
    _.Button = kHaierAcYrw02ButtonTempUp;
  if (fahrenheit) ...
```

Reply:
Sure, I will rework this code.